docs/contributing/999-rfcs/2023-10-01-workloads.md

```diff
@@ -0,0 +1,252 @@
+to support live updates which to instantly reload the program when the source changes.
+
+>
+> **Implementation Note:** Rebuilding the image every time will likely not be good enough and some
+> kind of live update solution will be needed. An example of such a mechanism can be found in
+> [Tilt](https://docs.tilt.dev/tutorial/5-live-update.html). Similar to Tilt's `live_update` clause,
+> this will likely require additional information such as where the code resides inside the
+> container and how to build it.
+>
+
+### Local debugging (P2)
+
+> Reqtag: `c:debugging`
+
+It should be possible to attach a debugger to a containerized workload (P2).
+
+### Building images in `wing compile` (P1)
+
+> Reqtag: `c:building`
+
+When referencing a local container image, `wing compile` should take care of building the image
+using `docker build`.
+
+> **Implementation Note:** Since `docker build` is potentially expensive, we should implement a
+> source hash-based mechanism to avoid building if the docker source context hasn't changed.
+
+### Publishing images during deployment (P1)
+
+> Reqtag: `c:publishing`
+
+During deployment, the image needs to be pushed into a container registry (e.g. Docker Hub or ECR or
+any other registry) and the latest image address wired up the to the IaC configuration.
+
+This flow can wildly vary for different platforms (e.g. some organizations might prefer to use a
+separate deployment system to publish images), but for the Terraform-based platforms (such as
+`tf-aws`), we should treat this like we treat publishing an AWS Lambda zip archives to an S3 bucket
+during deployment. Platform providers will need to explore a variety of configuration options to
+control this behavior (P2).
+
+We might be able to leverage the [Docker provider for
+Terraform](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs) with the approach
+[suggested here](https://stackoverflow.com/a/76608435) to push the image during deployment to a
+container registry.
+
+>
+> **Implementation Note:** Design the publishing mechanism with heavy hash-based caching in mind so
+> that both `wing compile` and the Terraform image publishing step will be no-op if the container
+> image hasn't changed.
+>
+
+### Container runtime environment
+
+> Reqtag: `c:environment`
+
+It should be possible to configure the runtime environment of the containers.
+
+The following is a non-exhaustive list of capabilities we will need to support:
+
+* Expose network ports of various types (http, http2, grpc, ??) (P1)
+* Command line arguments to pass to the container (P1)
+* Environment variables (P1)
+* Entrypoint command to override (if you wish to use a pre-built container image and just change the
+  entrypoint) (P1)
+* CPU and memory requirements (in some normalized model such as vCPUs, GBs) (P2)
+* GPUs (P2)
+* Liveness probe
+* Readiness probe (P1)
+* Sidecar containers (P2)
+* Volumes (P2)
+* Autoscaling (P1)
```

Comment:
I feel like we're conflating containers and container orchestrators. Most of the time this document mentions containers I feel like it's talking about pods. Am I thinking about this wrong?

Reply:
Moved autoscaling into a separate topic. I would map "workload" to a combination of "service + deployment", not "pod".